### PR TITLE
Implement form validation

### LINKED ISSUE #797: Show an error when trying to submit a form with missing/wrong data
**Is your feature request related to a problem? Please describe.**
When creating a lock with missing or wrong data, we do not let the user submit... (nothing happens), however I believe it is a good design  principal to *always* have something happening on user explicit actions (like clicking on a button). We should probably show an error with "Please make sure you complete all from fields" for now.

### PR DESCRIPTION
**Pull request by cellog**
# Description

Fixes #797 

This PR is a master PR, and may need to be split into a couple of PRs. However, I am unsure how best to do this, so that will be part of the review.

screen shot 2019-01-15 at 12 12 26 pm

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [X] If my code involves visual changes, I am adding applicable screenshots to this thread

PS: [Read how to write the perfect pull request](https://blog.github.com/2015-01-21-how-to-write-the-perfect-pull-request/)

### CODE REVIEW

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 1]
```diff
@@ -14,86 +14,157 @@ import {
...
     super(props, context)
     this.state = {
-      expirationDuration: 30,
-      expirationDurationUnit: 86400, // Days
-      keyPrice: '0.01',
-      keyPriceCurrency: 'ether',
-      maxNumberOfKeys: 10,
-      unlimitedKeys: false,
-      name: 'New Lock',
+      expirationDuration: props.expirationDuration,
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: This part can be in its own PR.

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 2]
```diff
@@ -14,86 +14,157 @@ import {
...
-    let { unlimitedKeys } = this.state
-    return (
-      (name === 'name' && typeof value === 'string' && value.length > 0) ||
-      ((name === 'expirationDuration' || name === 'maxNumberOfKeys') &&
-        !isNaN(value) &&
-        value > 0) ||
-      (name === 'maxNumberOfKeys' && value === '∞' && unlimitedKeys) ||
-      (name === 'keyPrice' && !isNaN(value) && value >= 0)
-    )
+    const missing = val => val === undefined || val === null || val === ''
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]:
```suggestion
    const missing = val => (val === undefined || val === null || val === '')
```

Comment (julien51) [reply unlock-app/src/components/creator/CreatorLockForm.js]: We can probably extract all of these functions from this one. Either at the top of the file or (even better), in helper file (`validations.js`?) which can be done in its own PR.

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 3]
```diff
@@ -14,86 +14,157 @@ import {
...
+        if (missing(value)) {
+          return FORM_LOCK_NAME_MISSING
+        }
+        break
+      case 'expirationDuration':
+      case 'maxNumberOfKeys': {
+        const error = {
+          expirationDuration: FORM_EXPIRATION_DURATION_INVALID,
+          maxNumberOfKeys: FORM_MAX_KEYS_INVALID,
+        }[name]
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: I see what you're trying to do here by not using `break` but I would recommend that you either use `break` and duplicate some code, or that you use an if statement. This would make the code easier to read and less likely to confuse people.

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 4]
```diff
@@ -14,86 +14,157 @@ import {
...
+      errors.push(invalidError)
+      return fieldValidity
+    }, {})
+
+    // the form can be submitted if and only if there are no errors triggered by any field
+    validityState.formValid = errors.length === 0
+    return { validityState, errors }
   }
 
   validate(name, value) {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: I like that refactor of `validate` which can probably be done in its own PR ;)

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 5]
```diff
@@ -14,86 +14,157 @@ import {
...
+      [name]: value,
+      valid: this.valid({ ...state.valid, [name]: value }),
+    }))
   }
 
   handleSubmit() {
-    if (document.querySelector('[data-valid="false"]')) return false
+    this.setState(state => {
+      const { validityState: valid, errors } = this.valid(state)
+      const { setError, resetError } = this.props
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: Pull that out of the `setState` callback (I think we should do the assignments at the beginning of the function)

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 6]
```diff
@@ -14,86 +14,157 @@ import {
...
-    const {
-      expirationDuration,
-      expirationDurationUnit,
-      keyPriceCurrency,
-      maxNumberOfKeys,
-      unlimitedKeys,
-      keyPrice,
-      name,
-    } = this.state
+      const { account, createLock, hideAction } = this.props
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: Combine with `const { setError, resetError } = this.props`?

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 7]
```diff
@@ -14,86 +14,157 @@ import {
...
+      const {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: Maybe move the part which does create the lock from `state` to its own function to make `handleSubmit` easier to read?

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 8]
```diff
@@ -119,24 +191,30 @@ export class CreatorLockForm extends React.Component {
             name="name"
             onChange={this.handleChange}
             defaultValue={name}
+            data-valid={valid.name}
           />
         </FormLockName>
         <FormLockDuration>
           <input
-            type="text"
+            type="number"
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: This is good! We could start with a PR which does just that for example (and does not change any of the logic)

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 9]
```diff
@@ -169,19 +250,34 @@ CreatorLockForm.propTypes = {
...
+  expirationDurationUnit: PropTypes.number,
+  keyPrice: PropTypes.string,
+  keyPriceCurrency: PropTypes.string,
+  maxNumberOfKeys: PropTypes.oneOfType([PropTypes.number, PropTypes.string]), // string is for '∞'
+  unlimitedKeys: PropTypes.bool,
+  name: PropTypes.string,
 }
 
-CreatorLockForm.defaultProps = {}
+CreatorLockForm.defaultProps = {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: Same as above: this is its own PR ;)

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 10]
```diff
@@ -195,15 +291,30 @@ const LockLabelUnlimited = styled(LockLabel)`
...
 
 const FormLockRow = styled(LockRow)`
-  ${LockRowGrid} input[type='text'] {
-    background-color: var(--lightgrey);
-    border: 0;
-    padding: 5px;
-    font-family: 'IBM Plex Sans', sans-serif;
-    font-size: 13px;
+  input[type='number'] {
+    -moz-appearance: textfield;
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: Meh. Are you sure this is a good idea to use prefixed-CSS stuff? I thought this would be handled by our build tools?

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 11]
```diff
@@ -219,22 +330,25 @@ const FormLockName = styled(LockName)`
 `
 
 const FormLockDuration = styled(LockDuration)`
-  input[type='text'] {
+  input[type='text'],
+  input[type='number'] {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: This stuff belongs in the PR which just set the field types ;)

File: unlock-app/src/components/helpers/ErrorMessage.js [hunk 12]
```diff
@@ -23,6 +23,17 @@ const defaultMessages = {
   NON_DEPLOYED_CONTRACT: (
     <p>Unlock has not been deployed on the current network.</p>
   ),
+
+  FORM_LOCK_NAME_MISSING: (
```
Comment (julien51) [inline unlock-app/src/components/helpers/ErrorMessage.js]: This could be in its own PR, along with the stuff in `src/errors.js`

File: unlock-app/src/components/creator/CreatorLockForm.js (hunk 10 above)
Comment (cellog) [reply unlock-app/src/components/creator/CreatorLockForm.js]: in this case, the appearance thing is only non-standard, and this is what was recommended to remove the up-down spinner for numbers on all browsers. Perhaps in a year or two, the browsers will standardize something, but for now, we are in 2010 when it comes to number input styling :)

File: unlock-app/src/components/creator/CreatorLockForm.js (hunk 5 above)
Comment (cellog) [reply unlock-app/src/components/creator/CreatorLockForm.js]: the only issue is that setState callbacks are async, I am not sure if we can rely on props in handleSubmit being the same when we finish. Although it may be OK, because it doesn't save any processor cycles or add other benefits, I don't think we should take this risk. What do you think?

File: unlock-app/src/components/creator/CreatorLockForm.js (hunk 3 above)
Comment (cellog) [reply unlock-app/src/components/creator/CreatorLockForm.js]: not only that, this allows us to use infinity in expirationDuration, which is probably not an intended feature :)

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 13]
```diff
@@ -36,31 +48,67 @@ export class CreatorLockForm extends React.Component {
...
+    // for each field, retrieve the error triggered by current state
+    // and then make sure we set it as existing.
+    const validityState = [
+      'expirationDuration',
+      //'expirationDurationUnit',
+      'keyPrice',
+      //'keyPriceCurrency',
+      'maxNumberOfKeys',
+      'name',
+    ].reduce((fieldValidity, field) => {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js]: nit: would `formValidity` be a better name to indicate that this is about all the fields?
